test/eredis_tcp_SUITE.erl

```diff
@@ -327,6 +341,132 @@ t_reconnect(Config) when is_list(Config) ->
+                        end),
+    process_flag(trap_exit, true),
+    Res = eredis:start_link("127.0.0.1", Port, [{database, 2},
+                                                {reconnect_sleep, no_reconnect}]),
+    ?assertMatch({error, {select_error, _}}, Res), %% einval or closed depending of timing
+    IsDead = receive {'EXIT', _, _} -> died
+             after 2000 -> still_alive end,
+    process_flag(trap_exit, false),
+    ?assertEqual(died, IsDead),
+    eredis_test_utils:stop_server(Pid).
+
+t_select_database_socket_closed2(Config) when is_list(Config) ->
+    %% Simulate that Redis closes the connection after the SELECT
+    %% command was sent.
+    {ok, Pid, Port} = eredis_test_utils:start_server(
+                        fun(ClientSocket) ->
+                                {ok, <<"SELECT 2\r\n">>} = gen_tcp:recv(ClientSocket, 0, 2000),
+                                ok = gen_tcp:close(ClientSocket)
+                        end),
+    process_flag(trap_exit, true),
+    Res = eredis:start_link("127.0.0.1", Port, [{database, 2},
+                                                {reconnect_sleep, no_reconnect}]),
+    ?assertMatch({error, {select_error, closed}}, Res),
+    IsDead = receive {'EXIT', _, _} -> died
+             after 2000 -> still_alive end,
+    process_flag(trap_exit, false),
+    ?assertEqual(died, IsDead),
+    eredis_test_utils:stop_server(Pid).
+
+t_send_error(Config) when is_list(Config) ->
+    %% Verify handling of send errors by using the send_timout_close option.
+    %% Connect an eredis client to a server and fill the clients sockets sendbuffer
+    %% until the socket is full. Then let eredis handle a query which will
+    %% trigger a send error (and reconnection attempts)
+    {ok, Pid, Port} = eredis_test_utils:start_server(
+                        fun(ClientSocket) ->
+                                inet:setopts(ClientSocket, [{recbuf, 512}])
+                        end),
+    {ok, C} = eredis:start_link("127.0.0.1", Port, [{reconnect_sleep, 100},
+                                                    {socket_options, [{send_timeout, 250},
+                                                                      {send_timeout_close, true},
+                                                                      {sndbuf, 512}]}]),
+    eredis_test_utils:await_connect(Pid),
+    [Socket] = get_tcp_ports(C),                %% Get eredis client socket
+    {error, timeout} = send_data_loop(Socket),  %% Fill the socket
+    ?assertEqual({error, enotconn}, eredis:q(C, ["GET", foo])),
+    timer:sleep(200),
+    [Socket2] = get_tcp_ports(C),               %% Get new eredis client socket
+    ?assertNotEqual(Socket, Socket2),           %% Reconnection attempts with new socket
+    eredis_test_utils:stop_server(Pid).
+
+send_data_loop(Socket) ->
+    Data = << <<0>> || _ <- lists:seq(1, 1024) >>,
+    case gen_tcp:send(Socket, Data) of
+        ok -> send_data_loop(Socket);
+        Error -> Error
+    end.
```

Comment:
its actually called 42 times on my machine before a send error is generated.. I seems that the os sets what it likes. But when not changing the sndbuf/recbuf sizes it will be a lot lot larger. There might be some other buffer in some erlang-layer that can take data aswell, dont know